test/conformance/source/environment.cpp

```diff
@@ -176,4 +179,145 @@ void DevicesEnvironment::TearDown() {
+        IL << ".spv";
+    } else {
+        error = "Undefined IL version: " + IL_version;
+        return "";
+    }
+
+    return IL.str();
+}
+
+std::string KernelsEnvironment::getKernelSourcePath(const std::string &kernel_name, uint32_t device_index) {
+    std::stringstream path;
+    path << instance->getKernelDirectory();
+    // il_postfix = supported_IL(SPIRV-PTX-...) + IL_version + extension(.spv -
+    // .ptx - ....)
+    std::string il_postfix = getSupportedILPostfix(device_index);
+
+    if (il_postfix.empty()) {
+        error = "failed getting device supported IL";
+        return "";
+    }
+
+    path << "/" << kernel_name << il_postfix;
+
+    uint32_t address_bits;
+    auto device = instance->GetDevices()[device_index];
+    if (urDeviceGetInfo(device, UR_DEVICE_INFO_ADDRESS_BITS, sizeof(uint32_t),
+                        &address_bits, nullptr)) {
+        error = "failed getting device address bits supported";
+        return "";
+    }
+    path << address_bits;
+
+    return path.str();
+}
+
+KernelsEnvironment::KernelSource KernelsEnvironment::LoadSource(const std::string &kernel_name, uint32_t device_index) {
+    std::string source_path =
+        instance->getKernelSourcePath(kernel_name, device_index);
+
+    if (source_path.empty()) {
+        error = "failed retrieving kernel source path for kernel: " + kernel_name;
+        return KernelSource{&kernel_name[0], nullptr, 0,
+                            UR_RESULT_ERROR_INVALID_BINARY};
+    }
+
+    if (cached_kernels.find(source_path) != cached_kernels.end()) {
+        return cached_kernels[source_path];
+    }
+
+    std::ifstream source_file;
+    source_file.open(source_path, std::ios::binary | std::ios::in | std::ios::ate);
+
+    if (!source_file.is_open()) {
+        error = "failed opening kernel path: " + source_path;
+        return KernelSource{&kernel_name[0], nullptr, 0,
+                            UR_RESULT_ERROR_INVALID_BINARY};
+    }
+
+    uint32_t source_size = static_cast<uint32_t>(source_file.tellg());
+    source_file.seekg(0, std::ios::beg);
+
+    char *source = new char[source_size];
+    source_file.read(source, source_size);
+    if (!source_file) {
+        source_file.close();
+        delete[] source;
+        error = "failed reading kernel source data from file: " + source_path;
+        return KernelSource{&kernel_name[0], nullptr, 0,
+                            UR_RESULT_ERROR_INVALID_BINARY};
+    }
+    source_file.close();
+
+    KernelSource kernel_source =
+        KernelSource{&kernel_name[0], reinterpret_cast<uint32_t *>(source), source_size, UR_RESULT_SUCCESS};
+
+    return cached_kernels[source_path] = kernel_source;
+}
+
+void KernelsEnvironment::SetUp() {
+    DevicesEnvironment::SetUp();
+    if (!error.empty()) {
+        FAIL() << error;
+    }
+}
+
+void KernelsEnvironment::TearDown() {
+    DevicesEnvironment::TearDown();
+    for (const auto &kernel : cached_kernels) {
+        delete[] kernel.second.source;
```

Comment:
Could maybe add a destructor to `KernelSource` to delete the source array then you just need to call `cached_kernels.clear()` which I think should call the destructor for each element.